### PR TITLE
dev/core#2559 Support tenancy in OAuth providers

### PR DESCRIPTION
**Pull request by JKingsnorth**
See https://lab.civicrm.org/dev/core/-/issues/2559

### CODE REVIEW

**Comment by MegaphoneJon** [comment]
@JKingsnorth test failures appear to be related.

**Comment by MegaphoneJon** [comment]
It looks like mostly undefined array key warnings.  I'm also reviewing the code, and will do an `r-run` if possible.  I have a client who needs this but because of red tape I don't have direct access to Azure, so I have to schedule time to screenshare with someone who does.

**Comment by JKingsnorth** [comment]
Yes, should be a quick fix for the tests, my local rig wasn't working so I let the test bot at it. I'll fix em up shortly :)

The test should just be: after updating and running the extension updates, you can create/edit an ms-exchange provider oauth client, and enter the tenant id.

I plan to update the documentation if the PR is accepted.

File: ext/oauth-client/CRM/OAuth/Upgrader.php [hunk 1]
```diff
@@ -36,4 +36,16 @@ public function upgrade_0001(): bool {
...
 
+  /**
+   * Add support for tenancy
+   *
+   * @return bool
+   * @throws Exception
+   */
+  public function upgrade_5581(): bool {
+    $this->ctx->log->info('Applying oauth-client update 5581. Adding tenant column to the civicrm_oauth_client table.');
+    CRM_Upgrade_Incremental_Base::addColumn($this->ctx, 'civicrm_oauth_client', 'tenant', 'varchar(128) NULL COMMENT "Tenant ID" AFTER guid');
```
Comment (seamuslee001) [inline ext/oauth-client/CRM/OAuth/Upgrader.php:47]: @JKingsnorth I believe you can change this to be
```suggestion
     $this->addTask('Add Tenant ID to OauthClient', 'addColumn', 'civicrm_oauth_client', 'tenant', 'varchar(128) NULL COMMENT "Tenant ID" AFTER guid');
```

Comment (JKingsnorth) [reply ext/oauth-client/CRM/OAuth/Upgrader.php]: Thanks @seamuslee001, that makes more sense. Done.

**Comment by JKingsnorth** [comment]
@MegaphoneJon / @seamuslee001 

I've altered the PR to add the logic into the CiviGenericProvider class, rather than creating a new one just for the ms-exchange. This should make it easier for other providers to add support for tenancy where required, whilst not causing bloat or breaking existing providers. But would be great to get an r-run test on sites that do use existing gmail providers [=

I've also added tests (ooh, ahh).

**Comment by JKingsnorth** [comment]
Squashed. Ready for any more review needed.

File: ext/oauth-client/CRM/OAuth/Upgrader.php [hunk 2]
```diff
@@ -36,4 +36,16 @@ public function upgrade_0001(): bool {
...
 
+  /**
+   * Add support for tenancy
+   *
+   * @return bool
+   * @throws Exception
+   */
+  public function upgrade_5581(): bool {
+    $this->ctx->log->info('Applying oauth-client update 5581. Adding tenant column to the civicrm_oauth_client table.');
+    $this->addTask('Add Tenant ID to civicrm_oauth_client', 'addColumn', 'civicrm_oauth_client', 'tenant', 'varchar(128) NULL COMMENT "Tenant ID" AFTER guid');
```
Comment (mattwire) [inline ext/oauth-client/CRM/OAuth/Upgrader.php:47]: Please can you wrap this line in a check so it's re-runnable:
```suggestion
  if (!CRM_Core_BAO_SchemaHandler::checkIfFieldExists('civicrm_oauth_client', 'tenant')) {
    $this->addTask('Add Tenant ID to civicrm_oauth_client', 'addColumn', 'civicrm_oauth_client', 'tenant', 'varchar(128) NULL COMMENT "Tenant ID" AFTER guid');
  }
```

Comment (JKingsnorth) [reply ext/oauth-client/CRM/OAuth/Upgrader.php:47]: @mattwire thanks for the eyes on this. I'm pretty sure the 'addColumn' task is already wrapped in the CRM_Core_BAO_SchemaHandler::checkIfFieldExists, if you look at: \CRM_Upgrade_Incremental_Base::addColumn [=

Comment (mattwire) [reply ext/oauth-client/CRM/OAuth/Upgrader.php:47]: Oh yes you are right!

**Comment by mattwire** [comment]
I've applied this patch on a ms-exchange site where tenant ID is required and it works. Also applied on a gmail site and it carries on working

**Comment by JKingsnorth** [comment]
Thanks @mattwire , and thanks for the review and merge.

**Pull request merged**